tap-version-14-specification.md

```diff
@@ -0,0 +1,741 @@
+would be overly difficult.
+
+## DIRECTIVES
+
+Directives are special notes that follow a # on the test line. Only
+two are currently defined: TODO and SKIP. Note that these two keywords
+are not case-sensitive.
+
+### TODO tests
+
+If the directive starts with `# TODO`, the test is counted as a todo
+test, and the text after TODO is the explanation.
+
+```
+not ok 13 # TODO bend space and time
+```
+
+Note that if the TODO has an explanation it must be separated from
+TODO by a space.  These tests represent a feature to be implemented or
+a bug to be fixed and act as something of an executable "things to do"
+list. They are not expected to succeed. Should a todo test point begin
+succeeding, the harness should report it as a bonus. This indicates
+that whatever you were supposed to do has been done and you should
+promote this to a normal test point.
+
+### Skipping tests
+
+If the directive starts with `# SKIP`, the test is counted as having
+been skipped. If the whole test file succeeds, the count of skipped
+tests is included in the generated output. The harness should report
+the text after # SKIP\S*\s+ as a reason for skipping.
+
+```
+ok 23 # skip Insufficient flogiston pressure.
+```
+
+Similarly, one can include an explanation in a Plan line, emitted if
+the test file is skipped completely:
+
+```
+1..0 # SKIP WWW::Mechanize not installed
+```
+
+## SUBTESTS
+
+A Subtest contains a block of test points grouped together.  This is
+useful for the output of test runners that collect several child
+processes' TAP streams, and output the result in TAP for another
+harness to interpret, or for tests with a predictable number of
+blocks, but where each block might not have a predictable number of
+tests.
+
+A Subtest is a TAP14 stream indented 4 spaces.  It has the following
+format:
+
+- A comment line with the name of the subtest: `/^ {4}# Subtest: (.*)$/`
+- A TAP14 stream, indented 4 spaces.
+- A (non-indented) test point indicating the success or failure of the
+  subtest, with a description matching the Subtest's name.
+
+For example:
+
+```
+TAP version 14
+1..2
+ok 1 - About to do a subtest
+    # Subtest: this is the subtest
+    ok 1 - This is ok
+    ok 2 - I am ok with how things are proceeding
+    not ok 3 - infinite loops # TODO Solve halting problem
+    not ok 4 - this is just a failure
+    1..4
+not ok 2 - this is the subtest
+  ---
+  plan: 4
+  count: 4
+  passing: 2
+  todo: 1
+  failed: 1
+  exitCode: 1
+  ...
+```
+
+If not bailed out, a Subtest MUST be introduced by the indented
+`# Subtest: ...` comment, and MUST be followed with a test point whose
+description matches the Subtest name.  If the Subtest is not
+successful, then the Subtest test point MUST be a `not ok` test point.
+
+A `Bail out!` in a Subtest SHOULD be immediately followed by a
+`Bail out!` in the parent test.
+
+If a Subtest TAP stream does not include a version number, it MUST be
+interpreted as a TAP14 stream anyway.  This is a backwards-compatible
```

Comment:
Yes, the context of this spec is to codify how we're extending TAP13 in practice.

The wording of this is tricky.  The issue here is that you don't want subtests to be interpreted as a TAP12 stream (since they'll likely have subtests and yaml diags), but including a `TAP version XX` line in a subtest upsets some existing parsers that don't require that the version start on column 0, so it must be acceptable to omit the version designator.

Perhaps it's better to say that, in absence of a `TAP version XX` in the subtest, it's assumed to be the same version as the parent test?  (Ie, 14+)